### DISCUSSION TITLE
SwiftUI cycle detected through attribute when binding to @Query parameter with view in if statement

### DISCUSSION THREAD

**Discussion by ethankay** [groue/GRDBQuery#5] [Q&A] [answered]
Building off of the same model [here](https://github.com/groue/GRDB.swift/discussions/1131), I have the following query:

```
import Combine
import GRDB
import GRDBQuery
import SwiftUI

struct WordRequest: Queryable {
  enum Ordering: CaseIterable {
    case byName
    case byAdded
  }
  
  var ordering: Ordering
  var filter: String = ""
  
  static var defaultValue: [Word] { [] }
  
  func publisher(in db: WordDb) -> AnyPublisher<[Word], Error> {
    ValueObservation
      .tracking(fetchValue(_:))
      .publisher(in: db.dbReader, scheduling: .immediate)
      .eraseToAnyPublisher()
  }
  
  func fetchValue(_ db: Database) throws -> [Word] {
    switch ordering {
    case .byName:
      return try Word.all().orderedByName().startingWith(filter).fetchAll(db)
    case .byAdded:
      return try Word.all().orderedByAdded().startingWith(filter).fetchAll(db)
    }
  }
}


extension DerivableRequest where RowDecoder == Word {
  func startingWith(_ word: String) -> Self {
    return filter(Word.Columns.name.like("\(word)%"))
  }
}
```

I then attach this into a SwiftUI like so:

```
import SwiftUI
import GRDBQuery

struct WordListView: View {
  @Query(WordRequest(ordering: .byAdded)) private var words: [Word]
  
  var body: some View {
    List() {
      ForEach(words, id: \.self) { word in
        Text(word.name)
      }
    }
    .toolbar {
      if !words.isEmpty {  // (1)
         // (2) Anything that edits and binds to "$words.ordering"
      }
    }
     .searchable(text: $words.filter)
  }
}
```

In the line above if I have (1) and (2) present then the app will crash whenever `$words` goes between empty and not empty with error: `=== AttributeGraph: cycle detected through attribute 364088 ===`. Is there some way around this? I'd like to have the button hide whenever the list of words is empty. Clearly having the binding on a `@Query` value that can go in and out with an if statement in a view causes this cycle detection problem. Thanks in advance!

**Comment by groue** [comment, marked as answer]
Hello @ethankay,

Please provide an example of such a "(2)".

**Reply by ethankay** [reply]
Of course, here's the specific code I have in this case:

```
private struct ToggleOrderingButton: View {
  @Binding var ordering: WordRequest.Ordering
  
  var body: some View {
    Menu {
      buttonForOrder(.byAdded)
      buttonForOrder(.byName)
    } label: {
      Image(systemName: "arrow.up.arrow.down.circle")
    }
  }
  
  private func buttonForOrder(_ order: WordRequest.Ordering) -> some View {
    Button {
      ordering = order
    } label: {
      Text(order.label)
      if ordering == order {
        Spacer()
        Image(systemName: "checkmark")
      }
    }
  }
}
```

Which then replaces (2) like so:
```
ToggleOrderingButton(ordering: $words.ordering)
```

**Reply by groue** [reply]
It looks like you need to use another SwiftUI technique, then. For example, did you try to [conditionally](https://www.avanderlee.com/swiftui/conditional-view-modifier/) apply the [hidden](https://developer.apple.com/documentation/swiftui/view/hidden()) modifier?

**Reply by ethankay** [reply]
Thanks for the suggestion @groue. Unfortunately that doesn't seem to work either. I replaced (1) and (2) with the extension from your link and then the following implementation:

```
ToggleOrderingButton(ordering: $words.ordering).if(!words.isEmpty) { view in
  view.hidden()
}
```

Same error: `=== AttributeGraph: cycle detected through attribute 473672 ===`

**Reply by groue** [reply]
😬 So you are facing the problems described there: https://www.objc.io/blog/2021/08/24/conditional-view-modifiers/, and yet another SwiftUI technique is needed. A classic SwiftUI experience, isn't it? It is cool until it breaks horribly, leaving us clueless. Please report the results of your investigations. I'll look for a solution as well.

**Reply by ethankay** [reply]
Such is life :)

Thanks @groue for your help!

**Reply by groue** [reply]
I could successfully display a conditional item that uses a `@Query` binding in the [GRDB demo app](https://github.com/groue/GRDB.swift/blob/v5.18.0/Documentation/DemoApps/GRDBCombineDemo/GRDBCombineDemo/Views/AppView.swift):

```swift
import GRDBQuery
import SwiftUI

struct AppView: View {
    @Query(PlayerRequest(ordering: .byScore)) private var players: [Player]

    var body: some View {
        NavigationView {
            PlayerList(players: players)
                .navigationBarTitle(Text("\(players.count) Players"))
                .toolbar { toolbarContent }
        }
    }
    
    @ToolbarContentBuilder private var toolbarContent: some ToolbarContent {
        ToolbarItem(placement: .navigationBarLeading) { ... }

        // Only display the ordering button if the list is not empty
        ToolbarItem(placement: .navigationBarTrailing) {
            Group {
                if !players.isEmpty {
                    ToggleOrderingButton(ordering: $players.ordering)
                }
            }
        }

        ...
    }
}
```

I could actually not reproduce your issue. I hope you will find inspiration from this example.
